pre_commit/languages/download.py

```diff
@@ -0,0 +1,267 @@
+    with envcontext(get_env_patch(envdir)):
+        yield
+
+
+class Platform:
+    OS: Final[MappingProxyType[str, str]] = MappingProxyType({
+        'Linux': 'linux',
+        'Darwin': 'darwin',
+        'Windows': 'windows',
+        'DragonFly': 'dragonfly',
+        'FreeBSD': 'freebsd',
+    })
+    CPU: Final[MappingProxyType[str, str]] = MappingProxyType({
+        'aarch64': 'arm64',
+        'aarch64_be': 'arm64be',
+        'arm': 'arm',
+        'i386': '386',
+        'i686': '386',
+        'x86_64': 'amd64',
+        'AMD64': 'amd64',
+        'ppc': 'ppc',
+        'ppc64': 'ppc64',
+        'ppc64le': 'ppc64le',
+    })
+
+    def __init__(self, value: str) -> None:
+        self.__value = value
+        os, cpu = self.parts
+        if os not in Platform.OS.values():
+            raise ValueError(f'invalid operating system `{os}`')
+        if cpu not in Platform.CPU.values():
+            raise ValueError(f'invalid CPU `{cpu}`')
+
+    @property
+    def parts(self) -> tuple[str, str]:
+        first, second = self.value.split('/', 1)
+        return (first, second)
+
+    @property
+    def os(self) -> str:
+        os, _ = self.parts
+        return os
+
+    @property
+    def cpu(self) -> str:
+        _, cpu = self.parts
+        return cpu
+
+    @property
+    def value(self) -> str:
+        return self.__value
+
+    @staticmethod
+    def host() -> Platform:
+        os = Platform.OS[platform.system()]
+        cpu = Platform.CPU[platform.machine()]
+        return Platform(f'{os}/{cpu}')
+
+    def __str__(self) -> str:
+        return f'{self.os}/{self.cpu}'
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, Platform):
+            return NotImplemented
+        return other.os == self.os and other.cpu == self.cpu
+
+
+class ChecksumError(ValueError):
+    def __init__(self, sri: SRI, message: str = 'Mismatched checksum'):
+        self.__sri = sri
+        self.__message = message
+
+    @property
+    def sri(self) -> SRI:
+        return self.__sri
+
+    @property
+    def message(self) -> str:
+        return self.__message
+
+    def __str__(self) -> str:
+        return f'''{self.__message}:
+ - expected: {self.sri.checksum}
+ - actual  : {self.sri.digest}'''
+
+
+class SRI:
+    def __init__(self, value: str):
+        self.__value = value
+        self._algorithm, self._checksum = self.value.split('-', 1)
+        if self.algorithm not in hashlib.algorithms_available:
+            raise ValueError('Invalid hashing algorithm.')
+
+        if b64encode(b64decode(self.checksum)).decode('utf-8') !=\
+                self.checksum:
+            raise ValueError('Invalid checksum string.')
+
+        self.__hasher = hashlib.new(self.algorithm)
+
+        if len(b64decode(self.checksum)) != self.__hasher.digest_size:
+            raise ValueError(
+                f'Invalid checksum string length for {self.algorithm}',
+            )
```

Comment:
It would make sense to include relevant context into the error message, so that I can easily understand why the error occured:

```suggestion
            raise ValueError(
                f'Invalid checksum string length of {...} for {self.algorithm}, expected {...}: {self.checksum}',
            )
```